src/ui/wxWidgets/AddEditPropSheetDlg.cpp

```diff
@@ -2656,16 +2668,40 @@ void AddEditPropSheetDlg::OnExpRadiobuttonSelected( wxCommandEvent& evt )
   bool Never = (evt.GetEventObject() == m_DatesTimesNeverExpireCtrl);
 
   if (Never) {
-    m_ExpirationTime = _("Never");
-    m_CurrentExpirationTime.Clear();
-    m_tttExpirationTime = time_t(0);
-    m_ExpirationTimeInterval = 90;
-    wxDateTime xdt(wxDateTime::Now());
+    //m_tttExpirationTime = time_t(0);
+    // If it has an interval, keep it.  Otherwise, use the user defined default.
+    // This applies to new entries and old entries with no interval setting.
+    //
+    // Some entries, created with recent versions of pwsafe, will have an interval but no date,
+    // which is interpreted as "Never".  But, we need to use the given interval to prevent
+    // a false change detection when the entry is opened and closed.  If an actual change is made,
+    // the entry will be re-written correctly.
+    if (!m_ExpirationTimeInterval) { // Compatibility
+      m_ExpirationTimeInterval = PWSprefs::GetInstance()->GetPref(PWSprefs::DefaultExpiryDays);;
+    }
+    // Needed here for initialization and the above compatibility problem.
+    wxDateTime xdt(wxDateTime::Today());  // Today returns time part == 0
     xdt += wxDateSpan(0, 0, 0, m_ExpirationTimeInterval);
     m_DatesTimesExpiryDateCtrl->SetValue(xdt);
+    m_tttExpirationTime = xdt.GetTicks();
+    m_Recurring = (m_Type == SheetType::ADD) ? true : false;
+
+  } else if (On) {  // Specific Date
+    if (!m_ExpirationTimeInterval) { // Compatibility
+      m_ExpirationTimeInterval = PWSprefs::GetInstance()->GetPref(PWSprefs::DefaultExpiryDays);;
+    }
     m_Recurring = false;
-    TransferDataToWindow();
+
+  } else {  // Interval
+    wxDateTime xdt(wxDateTime::Today());  // Today returns time part 0
+    xdt += wxDateSpan(0, 0, 0, m_DatesTimesExpiryTimeCtrl->GetValue());
+    m_DatesTimesExpiryDateCtrl->SetValue(xdt);
+    m_tttExpirationTime = xdt.GetTicks();
+    m_Recurring = true;
   }
+  // In add mode, the item data is not prepared for this, fields will go blank.
```

Comment:
This seems like a tight coupling between this event handler and the control, in that we're making assumptions about another control's behavior. Instead, could we still send the data, but implement proper handling logic in the control? I don't know how we're populating the field there, so this may not make sense.

Reply:
I think the problem is that the function that initializes the dialog controls (it was called UpdateExpTimes, I changed it to InitializeExpTimes because it's only called once) finishes by calling OnExpRadiobuttonSelected() to complete the initialization.  It worked okay when editing an existing entry, but caused problems when filling-in a new entry.  I thought it was strange to call an event handler like that, but I didn't change it, I just tried to make it work.  

I'm going to re-think that decision and see if I can refactor that code for a cleaner separation.